[JENKINS-73812] Minimum bind password length for FIPS compliance

nevingeorgesunny opened the pull request:
In the domain section of active directory
if you enter a short password , and the focus leaves from the field we get this message
<img width="1054" alt="image" src="https://github.com/user-attachments/assets/88897fd6-f0e3-444c-b907-816ed7358d4a">

is the user still clicks on on test domain it thows angry Jenkins error 

<img width="1054" alt="image" src="https://github.com/user-attachments/assets/5931cd15-7a89-4a74-ac03-aa6fb0f13326">

and if the user clicks on save , this same error appears
<img width="1054" alt="image" src="https://github.com/user-attachments/assets/6e0c10fc-83a1-48a5-a00d-7116abd8220b"> 

if the user click on apply a popup with error shows
<img width="1054" alt="image" src="https://github.com/user-attachments/assets/1c81acf6-0197-4470-bb7e-b9394943a14b">

<img width="1054" alt="image" src="https://github.com/user-attachments/assets/8a5bcd7a-2f45-422b-ae6b-7db2c92baea9">

PereBueno commented on src/main/java/hudson/plugins/active_directory/ActiveDirectoryDomain.java:
```diff
@@ -266,6 +272,17 @@ public ListBoxModel doFillTlsConfigurationItems() {
+         * Displays an error message if the provided password is less than 14 characters
+         * while in FIPS mode. This message is triggered when the bindPassword field loses focus.
+         */
+        public FormValidation doCheckBindPassword(@QueryParameter String bindPassword) {
```
```suggestion
        @RequirePOST
        public FormValidation doCheckBindPassword(@QueryParameter String bindPassword) {
```

nevingeorgesunny replied:
have added @RequirePOST

PereBueno commented on src/main/java/hudson/plugins/active_directory/ActiveDirectoryDomain.java:
```diff
@@ -266,6 +272,17 @@ public ListBoxModel doFillTlsConfigurationItems() {
+         * while in FIPS mode. This message is triggered when the bindPassword field loses focus.
+         */
+        public FormValidation doCheckBindPassword(@QueryParameter String bindPassword) {
+            if(FIPS140.useCompliantAlgorithms() && StringUtils.length(bindPassword) < 14) {
```
```suggestion
            Jenkins.get().checkPermission(Jenkins.ADMINISTER);
            if(FIPS140.useCompliantAlgorithms() && StringUtils.length(bindPassword) < 14) {
```

fcojfernandez replied:
No, please. Check MANAGE isntead to prevent a permissions error in case someone with MANAGE enters the page
```suggestion
            Jenkins.get().checkPermission(Jenkins.MANAGE);
            if(FIPS140.useCompliantAlgorithms() && StringUtils.length(bindPassword) < 14) {
```

fcojfernandez replied:
btw, MANAGE not supported in this plugin: https://github.com/jenkinsci/active-directory-plugin/pull/201/files#r1780770012

olamy replied:
checking permission should be done for expensive check or checks including sending credentials (i.e possible leaks).
here the code is just checking length of a password so not sure it's really mandatory.

nevingeorgesunny replied:
we can mark this as resolved right ?as we will not be checking permission

PereBueno replied:
Well, that's a security warn, I agree it's not that important here, so if code can be merged with that ok for me

olamy replied:
correct it's a warning which can be safely ignored

PereBueno commented on src/main/java/hudson/plugins/active_directory/ActiveDirectoryDomain.java:
When performing the "Test Domain" validation we're just showing an angry jenkins with no further information.
We would have a quite better UX if we added the check in https://github.com/jenkinsci/active-directory-plugin/blob/master/src/main/java/hudson/plugins/active_directory/ActiveDirectoryDomain.java#L312 and simply return a `FormValidation.error()`

fcojfernandez replied:
you mean that in `doValidateTest` if the IllegalArgumentExcetpion is thrown, to catch it and return instead a `FormValidation.error()`, right? If so, I think that's a good idea

PereBueno replied:
yes

nevingeorgesunny replied:
I Have tried this but the problem is , when I click on "Test Domain". it invokes 

```
@DataBoundConstructor
    public ActiveDirectoryDomain(String name, String servers, String site, String bindName, String bindPassword, TlsConfiguration tlsConfiguration) {
        this.name = name;
        // Gives exception if Password is set lees than 14 chars long in FIPS mode.
        if(FIPS140.useCompliantAlgorithms() && StringUtils.length(bindPassword) < 14) {
            throw new IllegalArgumentException(Messages.passwordTooShortFIPS());
        }

        // Append default port if not specified
        servers = fixEmpty(servers);
        if (servers != null) {
            String[] serversArray = servers.split(",");
            for (int i = 0; i < serversArray.length; i++) {
                if (!serversArray[i].contains(":")) {
                    serversArray[i] += ":3268";
                }
            }
            servers = StringUtils.join(serversArray, ",");
        }
        this.servers = servers;
        this.site = fixEmpty(site);
        this.bindName = fixEmpty(bindName);
        this.bindPassword = Secret.fromString(fixEmpty(bindPassword));
        this.tlsConfiguration = tlsConfiguration;
    }
```

correct me if im wrong . but I have to do the FIPS check in the method anyway because This  is also called when I click on save or apply .

so in this check when i throw `throw new IllegalArgumentException(Messages.passwordTooShortFIPS());`. We get the angry jenkins exception.

If we need to handel this more gracefully , I think we need to add some custom  error handelling I believe ?

fcojfernandez replied:
it's done as part of https://github.com/jenkinsci/active-directory-plugin/pull/201/files#diff-501cc4984dc677ba56eb812268a11cedbd9390137ed9b274e34a30a06adb72d8R290-R303 so once merged, you'll get if for free

fcojfernandez commented on src/test/java/hudson/plugins/active_directory/ActiveDirectoryDomainTest.java:
```diff
@@ -0,0 +1,66 @@
+	public void testPasswordTooShortInFIPSMode() {
+		// Create an instance of ActiveDirectoryDomain with a short password and assert exception
+		IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> {
+			new ActiveDirectoryDomain("example.com", "server", "site", "bindName", "short", TlsConfiguration.TRUST_ALL_CERTIFICATES);
```
Beware of using TRUST_ALL_CERTIFICATES. In FIPS mode it won't be allowed after https://github.com/jenkinsci/active-directory-plugin/pull/201 and the test will fail

nevingeorgesunny replied:
I have updated this in the latest push

olamy commented on src/main/java/hudson/plugins/active_directory/ActiveDirectoryDomain.java:
```diff
@@ -149,6 +150,11 @@ public ActiveDirectoryDomain(String name, String servers, String site, String bi
         this.name = name;
+        // Gives exception if Password is set lees than 14 chars long in FIPS mode.
+        if(FIPS140.useCompliantAlgorithms() && StringUtils.length(bindPassword) < 14) {
+            throw new IllegalArgumentException(Messages.passwordTooShortFIPS());
```
maybe better 
```
 throw new hudson.model.Descriptor.FormException(Messages.passwordTooShortFIPS(), "bindPassword");
```

see https://issues.jenkins.io/browse/JENKINS-73771

fcojfernandez replied:
problem with that is that the FormException wasn't available when the first FIPS PRs were opened and then there would be a mismatch between PRs. Previous PR were throwing IllegalArgumentException

olamy replied:
we did this here https://github.com/jenkinsci/credentials-plugin/pull/558 last week

rsandell replied:
I was thinking of using `FormException` as well in the Maile plugin fix. But it is a checked exception and I didn't want to declare it in the constructor signature. So `IllegalArgumentException` seemed better imo.

olamy replied:
fair enough, I can live with that :) 
but if not doing it now this will probably never happen though.

rsandell reviewed (changes requested):
I would like to see some more integration type tests. Testing that calling a constructor fails is all well an good, but is not very future proof when the class evolves. 
Testing that the validation error shows up on the configuration page and that it fails when the actual form is submitted is better.
Also a test for a bad JCasC configuration should also be done.

nevingeorgesunny commented:
> I would like to see some more integration type tests. Testing that calling a constructor fails is all well an good, but is not very future proof when the class evolves. Testing that the validation error shows up on the configuration page and that it fails when the actual form is submitted is better. Also a test for a bad JCasC configuration should also be done.
all the following test are run in FIPS mode

I have added these Integration test 
1. Using local data, small password is set , click on "test Domain" this will throw angry Jenkins error 
2. Using local data , small password is set , click on "Save"this will throw angry Jenkins error
3. Using local data , small password is set , click on "Apply"this will throw angry Jenkins error
4. Using local data , valid password is set , click on "Apply" this will return 200 then I update the password to small then i click on "Apply "this will throw angry Jenkins error

I have added some JCasC tests
1. Setting up a Jenkins with CasC YAML containing small password, this will throw error 
2. Setting up a Jenkins with CasC YAML containing valid password, this should work

fcojfernandez commented on src/test/java/hudson/plugins/active_directory/ActiveDirectoryDomainIntegrationTest.java:
```diff
@@ -0,0 +1,159 @@
+		getButtonByText(form, "Apply").click();
+	}
+
+	/**
+	 * Tests the behavior of the "Test Domain" button when a short password is configured.
+	 *
+	 * <p>For the preconfigured value, the password is "small" in the local data.
+	 * When the "Test Domain" button is clicked, the page should display an error message
+	 * indicating that the password is too short, along with an "angry Jenkins" error message.</p>
+	 *
+	 */
+//	@LocalData
+//	@Test
+//	public void testActiveDirectoryDomainTestDomainButtonClickWithShortPassword() throws Exception {
+//		JenkinsRule.WebClient webClient = jenkins.createWebClient();
+//		// Navigate to the configuration page
+//		HtmlPage configPage = webClient.goTo("configureSecurity");
+//		HtmlForm form = configPage.getFormByName("config");
+//
+//		//Check that the password is too short message is present
+//		assertTrue(form.asNormalizedText().contains(Messages.passwordTooShortFIPS()));
+//
+//		// Click the "Test Domain" button
+//		HtmlPage resultPage = getButtonByText(form, "Test Domain").click();
+//
+//		webClient.waitForBackgroundJavaScript(2000); // Wait for up to 5 seconds
+//
+//		String responseContent = resultPage.asNormalizedText();
+//		// Assert that the error message is present in the page content
+//		assertTrue(responseContent.contains("A problem occurred while processing the request"));
+//
+//		//Check that the password is too short message is present
+//		assertTrue(responseContent.contains(Messages.passwordTooShortFIPS()));
+//	}
```
why adding a commented test? please either fix it or remove it, but don't add a commented method

nevingeorgesunny replied:
I keep getting the error [Test Domain] not found , but this does not happen locally , am i configuring something wrong ?

rsandell commented on src/main/java/hudson/plugins/active_directory/ActiveDirectoryDomain.java:
```diff
@@ -266,6 +269,18 @@ public ListBoxModel doFillTlsConfigurationItems() {
+            }
+            return FormValidation.ok();
+        }
+
```
Missing a `readResolve` check as we've done for the other fixes.

fcojfernandez replied:
as stated in several places, `readResolve` is not even needed as FIPS doesn't support backward compatibility to begin with

rsandell replied:
There is already one such check for tlsCert in this plugin so we should try to be consistent.

fcojfernandez replied:
one check that I already explained it wasn't necessary and anyway you insisted to include making now needed that artificial consistency. Anyway, on my side I'm not converting this thread in a never ending discussion. The check was included yesterday and I'm resolving the conversation

rsandell commented on src/test/java/hudson/plugins/active_directory/ActiveDirectoryDomainIntegrationTest.java:
```diff
@@ -0,0 +1,171 @@
+	/**
+	 * Tests the behavior of the "Save" button when a short password is configured.
+	 *
+	 * <p>For the preconfigured value, the password is "small" in the local data.
```
So this test would break when adding the readResolve check.

nevingeorgesunny replied:
@fcojfernandez  had pushed some changes that fixed this.
I believe this can be marked as removed

rsandell commented on src/test/java/hudson/plugins/active_directory/ActiveDirectoryDomainIntegrationTest.java:
```diff
@@ -0,0 +1,171 @@
+
+import static hudson.Functions.isWindows;
+
+public class ActiveDirectoryDomainIntegrationTest {
```
These tests are explicitly for FiPS so it should be noted in the class name for easier future investigations.

rsandell commented on src/test/java/hudson/plugins/active_directory/ActiveDirectoryDomainTest.java:
```diff
@@ -0,0 +1,66 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThrows;
+
+public class ActiveDirectoryDomainTest {
```
These tests are explicitly for FiPS so it should be noted in the class name for easier future investigations.

fcojfernandez commented on src/test/java/hudson/plugins/active_directory/ActiveDirectoryDomainFIPSEnabledIntegrationTest.java:
```diff
@@ -0,0 +1,146 @@
+		webClient.waitForBackgroundJavaScript(5000);
+
+		form.getInputByName("_.bindPassword").setValue("short");
+		if(!isWindows()) {
+			form.getSelectByName("_.tlsConfiguration").setSelectedAttribute("JDK_TRUSTSTORE", true);
+		}
+
```
Because of `assumeFalse("JENKINS-73847", isWindows());` this can be omitted.
```suggestion
```

fcojfernandez commented on src/test/java/hudson/plugins/active_directory/ActiveDirectoryDomainFIPSEnabledIntegrationTest.java:
```diff
@@ -0,0 +1,146 @@
+		HtmlForm form = configPage.getFormByName("config");
+
+		form.getInputByName("_.bindPassword").setValue("short");
+		if(!isWindows()) {
+			form.getSelectByName("_.tlsConfiguration").setSelectedAttribute("JDK_TRUSTSTORE", true);
+		}
```
For clarity, let's say the JENKINS ticket
```suggestion
		if(!isWindows()) {
		    // JENKINS-73847
		    form.getSelectByName("_.tlsConfiguration").setSelectedAttribute("JDK_TRUSTSTORE", true);
		}
```

fcojfernandez reviewed (approved):
Minor comments. Please apply the suggestions
